lib/phoenix/code_reloader/server.ex

```diff
@@ -129,25 +138,37 @@ defmodule Phoenix.CodeReloader.Server do
     end
   end
 
-  defp mix_compile({:module, Mix.Task}, compilers) do
-    if Mix.Project.umbrella? do
-      Enum.each Mix.Dep.Umbrella.cached, fn dep ->
-        Mix.Dep.in_dependency(dep, fn _ ->
-          mix_compile_unless_stale_config(compilers)
-        end)
-      end
-    else
-      mix_compile_unless_stale_config(compilers)
-      :ok
-    end
+  defp mix_compile({:module, Mix.Task}, compilers, apps_to_reload) do
+    reloaded = mix_compile_deps(Mix.Dep.cached, apps_to_reload, compilers)
+
+    mix_compile_project(Mix.Project.get.project[:app], apps_to_reload, reloaded, compilers)
   end
-  defp mix_compile({:error, _reason}, _) do
+  defp mix_compile({:error, _reason}, _, _) do
     raise "the Code Reloader is enabled but Mix is not available. If you want to " <>
           "use the Code Reloader in production or inside an escript, you must add " <>
           ":mix to your applications list. Otherwise, you must disable code reloading " <>
           "in such environments"
   end
 
+  defp mix_compile_deps(deps, apps_to_reload, compilers) do
+    deps
+    |> Enum.filter(&(&1.app in apps_to_reload))
+    |> Enum.map(fn dep ->
+      Mix.Dep.in_dependency dep, fn _ ->
+        mix_compile_unless_stale_config(compilers)
+      end
+    end)
+  end
+
+  defp mix_compile_project(nil, _, _, _), do: :ok
```

Comment:
I think we can remove this clause because I don't think it can ever be nil. Plus `nil` is never supposed to be in `to_reload` or in `reloaded`.

Reply:
we can but then we need to wrap this function call in if clause. `Mix.Project.get.project[:app]` is nil in umbrella project, so it can be nil.

Reply:
and `Mix.Project.config()[:app]` behaves the same